lib/autoproj/cli/watch.rb

```diff
@@ -32,79 +56,51 @@ def callback
-                source_packages.each do |pkg_name|
-                    pkg = ws.manifest.find_autobuild_package(pkg_name)
-                    manifest_file = File.join(pkg.srcdir, 'manifest.xml')
+                @source_packages_dirs.each do |pkg_srcdir|
+                    next unless File.exist? pkg_srcdir
+                    create_dir_watcher(pkg_srcdir, included_paths: ["manifest.xml"])
 
-                    next unless File.exist? pkg.srcdir
-                    create_dir_watcher(pkg.srcdir, /^manifest.xml$/)
+                    manifest_file = File.join(pkg_srcdir, 'manifest.xml')
                     next unless File.exist? manifest_file
                     create_file_watcher(manifest_file)
                 end
             end
 
-            def create_autobuild_watchers(dir)
-                Dir[File.join(dir, '*.autobuild')].each do |file|
-                    create_file_watcher(file)
-                end
-            end
-
-            def create_manifests_watchers(dir)
-                Dir[File.join(dir, '**/*.xml')].each do |file|
-                    create_file_watcher(file)
-                end
-            end
-
-            def create_ruby_watchers(dir)
-                Dir[File.join(dir, '**/*.rb')].each do |file|
-                    create_file_watcher(file)
-                end
-            end
-
-            def create_pkg_set_watchers
-                ws.manifest.each_package_set do |pkg_set|
-                    pkg_set_dir = pkg_set.raw_local_dir
-                    create_dir_watcher(pkg_set_dir,
-                                       /^manifests$/,
-                                       /^*\.autobuild$/,
-                                       /^*\.rb$/)
-
-                    create_autobuild_watchers(pkg_set_dir)
-                    create_ruby_watchers(pkg_set_dir)
-
-                    manifests_dir = File.join(pkg_set_dir, 'manifests')
-                    next unless File.exist?(manifests_dir)
-                    create_manifests_watchers(manifests_dir)
-                    create_dir_watcher(manifests_dir, /^*\.xml$/,
-                                       recursive: true)
-                end
-            end
-
             def start_watchers
                 create_file_watcher(ws.config.path)
-                create_file_watcher(ws.manifest_file_path)
                 create_src_pkg_watchers
-                create_pkg_set_watchers
+                create_dir_watcher(ws.config_dir,
+                    excluded_paths: [/(^|#{File::SEPARATOR})\./],
+                    inotify_flags: [:recursive])
+                FileUtils.mkdir_p ws.remotes_dir
+                create_dir_watcher(ws.remotes_dir,
+                    excluded_paths: [/(^|#{File::SEPARATOR})\./],
+                    inotify_flags: [:recursive])
             end
 
             def cleanup_notifier
-                notifier.watchers.each_value(&:close)
+                notifier.watchers.dup.each_value(&:close)
```

Comment:
Because I'm assuming that closing a watcher on rb-inotify actually removes it from the `watchers` set (a.k.a. "a modify-while-iterating" bug in making). Being defensive here ...

Reply:
:+1: